src/expire.c

```diff
@@ -207,39 +241,31 @@ void activeExpireCycle(int type) {
      * 2) If last time we hit the time limit, we want to scan all DBs
      * in this iteration, as there is work to do in some DB and we don't want
      * expired keys to use memory for too much time. */
-    if (dbs_per_call > server.dbnum || timelimit_exit) dbs_per_call = server.dbnum;
+    if (dbs_per_call > server.dbnum || state->timelimit_exit) dbs_per_call = server.dbnum;
 
-    /* We can use at max 'config_cycle_slow_time_perc' percentage of CPU
-     * time per iteration. Since this function gets called with a frequency of
-     * server.hz times per second, the following is the max amount of
-     * microseconds we can spend in this function. */
-    timelimit = config_cycle_slow_time_perc * 1000000 / server.hz / 100;
-    timelimit_exit = 0;
-    if (timelimit <= 0) timelimit = 1;
+    state->timelimit_exit = false;
 
-    if (type == ACTIVE_EXPIRE_CYCLE_FAST) timelimit = config_cycle_fast_duration; /* in microseconds. */
+    if (cycleType == ACTIVE_EXPIRE_CYCLE_FAST) timelimit_us = config_cycle_fast_duration; /* in microseconds. */
 
     /* Accumulate some global stats as we expire keys, to have some idea
      * about the number of keys that are already logically expired, but still
      * existing inside the database. */
     long total_sampled = 0;
     long total_expired = 0;
 
-    /* Try to smoke-out bugs (server.also_propagate should be empty here) */
-    serverAssert(server.also_propagate.numops == 0);
-
     /* Stop iteration when one of the following conditions is met:
      *
      * 1) We have checked a sufficient number of databases with expiration time.
      * 2) The time limit has been exceeded.
      * 3) All databases have been traversed. */
-    for (j = 0; dbs_performed < dbs_per_call && timelimit_exit == 0 && j < server.dbnum; j++) {
+    for (j = 0; dbs_performed < dbs_per_call && state->timelimit_exit == 0 && j < server.dbnum; j++) {
         /* Scan callback data including expired and checked count per iteration. */
-        expireScanData data;
+        expireScanData data = {0};
         data.ttl_sum = 0;
         data.ttl_samples = 0;
+        data.max_entries = keys_per_loop * 4;
 
-        serverDb *db = server.db[(current_db % server.dbnum)];
+        serverDb *db = server.db[(state->current_db % server.dbnum)];
         data.db = db;
```

Comment:
How about adding a line `if (!db) continue;`?

This would eliminate 3 checks for `db` below.  I'd also increment the current_db here as well, eliminating the line below.  This prevents the possibility of a misplaced `continue` causing the db number to not increment.

```suggestion
        serverDb *db = server.db[(state->current_db++ % server.dbnum)];
        data.db = db;
        if (!db) continue;
```